Cleanup cached bundles after conduct load command

fsat opened the pull request:
Cleanup will be based on the name and the compat version of the recently loaded bundle.

The following will be kept:
- The recently loaded bundle.
- The latest version of the bundle in the cache having the same name and compat version as the recently loaded bundle.

All other versions of the bundle having the same name and compat version will be removed.

The latest version is decided by the timestamp of the bundle file on the disk.

fsat committed:
Cleanup cached bundles after conduct load command

Cleanup will be based on the name and the compat version of the recently loaded bundle.

The following will be kept:
- The recently loaded bundle.
- The latest version of the bundle in the cache having the same name and compat version as the recently loaded bundle.

All other versions of the bundle having the same name and compat version will be removed.

The latest version is decided by the timestamp of the bundle file on the disk.

fsat commented:
Marked as wip - I want to perform some manual test

fsat commented:
**Manual test setup**

Install CLI with local change.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ pip3 install -e .
Obtaining file:///Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli
Requirement already satisfied (use --upgrade to upgrade): requests>=2.6.0 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): requests-toolbelt>=0.7.0 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): argcomplete>=0.8.1 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pyhocon==0.2.1 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): arrow>=0.6.0 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): jsonschema==2.4 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pager<4.0,>=3.3 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): prettytable<1.0,>=0.7 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pygments<3.0,>=2.0 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): six<2.0,>=1.9 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): toml<1.0,>=0.9 in /usr/local/lib/python3.5/site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pyparsing==2.0.3 in /usr/local/lib/python3.5/site-packages (from pyhocon==0.2.1->conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): python-dateutil in /usr/local/lib/python3.5/site-packages (from arrow>=0.6.0->conductr-cli==0.45)
Installing collected packages: conductr-cli
  Running setup.py develop for conductr-cli
Successfully installed conductr-cli-0.45
You are using pip version 8.1.2, however version 9.0.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

fsat commented:
**Manual test with Conduct Load using V2 of the Control Protocol**

Start sandbox with ConductR 1.1 since it has both V1 and V2 Control Protocol.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ docker rm -f cond-0 cond-1 cond-2
cond-0
Error response from daemon: no such id: cond-1
Error response from daemon: no such id: cond-2
Error: failed to remove containers: [cond-1 cond-2]

Felixs-MBP-2:conductr-cli felixsatyaputra$ sandbox run 1.1.11
Starting ConductR nodes..
Starting container cond-0..
1ff8c56341540237ffdbcc42bde90921b922a2077f5aaf365126f57ac7203b4b
Waiting for ConductR to start...
ConductR has been started. Check current bundle status with: conduct info

Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct info
ID       NAME    #REP  #STR  #RUN
5a79de9  eslite     1     0     0
```

Prime the cache with older versions of the `cassandra` bundle.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ ls -lart ~/.conductr/cache/cassandra-v3-*
-rw-r--r--  1 felixsatyaputra  staff  44578327 22 Mar  2016 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-1f869c5d2b5d94ddf1b46c35f4d48bc58d24ddc46aeb66327ce23f2f9a44fc69.zip
-rw-r--r--  1 felixsatyaputra  staff  44578324  7 Sep 17:55 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-9e348c46cb7eee755f932240ca4c3d8fb3cb17ebaa8c18c02b47caa5d294a443.zip
-rw-------  1 felixsatyaputra  staff  44578329 21 Oct 13:45 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-067414825020a56df569e948de18e7010fb05534d20e1b749760033652a572cd.zip
-rw-------  1 felixsatyaputra  staff  44578324  2 Nov 16:46 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-6a2f3414f4f9de39c78556aeae21952768f5a1a570bd3f446b7982567219b7b4.zip
```

Load latest `cassandra` from bintray. Bundle will be downloaded from Bintray and loaded into ConductR.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct load cassandra
Retrieving bundle...
Bintray credentials loaded from /Users/felixsatyaputra/.bintray/.credentials
Loading bundle from cache typesafe/bundle/cassandra
Retrieving file:///Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/cassandra
Bintray credentials loaded from /Users/felixsatyaputra/.bintray/.credentials
Resolving bundle typesafe/bundle/cassandra
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/cassandra/v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
[#################################################] 100%
Loading bundle to ConductR...
[#################################################] 100%
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc waiting to be installed
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc installed
Bundle loaded.
Start bundle with: conduct run 4f71ca2
Unload bundle with: conduct unload 4f71ca2
Print ConductR info with: conduct info

Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct info
ID       NAME       #REP  #STR  #RUN
5a79de9  eslite        1     0     1
4f71ca2  cassandra     1     0     0
```

The older versions of Cassandra bundle is cleaned up. Only the latest version and the recently loaded Cassandra bundle is present in the cache.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ ls -lart ~/.conductr/cache/cassandra-v3-*
-rw-------  1 felixsatyaputra  staff  44578324  2 Nov 16:46 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-6a2f3414f4f9de39c78556aeae21952768f5a1a570bd3f446b7982567219b7b4.zip
-rw-------  1 felixsatyaputra  staff  44578324 14 Nov 17:58 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
```

Subsequent load works as well.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct load cassandra
Retrieving bundle...
Bintray credentials loaded from /Users/felixsatyaputra/.bintray/.credentials
Loading bundle from cache typesafe/bundle/cassandra
Retrieving from cache /Users/felixsatyaputra/.conductr/cache/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
Loading bundle to ConductR...
[#################################################] 100%
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc is installed
Bundle loaded.
Start bundle with: conduct run 4f71ca2
Unload bundle with: conduct unload 4f71ca2
Print ConductR info with: conduct info

Felixs-MBP-2:conductr-cli felixsatyaputra$ ls -lart ~/.conductr/cache/cassandra-v3-*
-rw-------  1 felixsatyaputra  staff  44578324  2 Nov 16:46 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-6a2f3414f4f9de39c78556aeae21952768f5a1a570bd3f446b7982567219b7b4.zip
-rw-------  1 felixsatyaputra  staff  44578324 14 Nov 17:58 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip

Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct info
ID       NAME       #REP  #STR  #RUN
5a79de9  eslite        1     0     1
4f71ca2  cassandra     1     0     0
```

fsat commented:
**Manual test with Conduct Load using V1 of the Control Protocol**

Start sandbox with ConductR 1.1 since it has both V1 and V2 Control Protocol.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ docker rm -f cond-0 cond-1 cond-2
cond-0
Error response from daemon: no such id: cond-1
Error response from daemon: no such id: cond-2
Error: failed to remove containers: [cond-1 cond-2]

Felixs-MBP-2:conductr-cli felixsatyaputra$ sandbox run 1.1.11
Starting ConductR nodes..
Starting container cond-0..
32b77f54b209ad09b7f7cf7f0c91e5049f19d535a6a6d55e1c82e2cb4337c423
Waiting for ConductR to start...
ConductR has been started. Check current bundle status with: conduct info

Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct info
ID       NAME    #REP  #STR  #RUN
5a79de9  eslite     1     0     0
```

Prime the cache with older versions of the `cassandra` bundle.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ ls -lart ~/.conductr/cache/cassandra-v3-*
-rw-r--r--  1 felixsatyaputra  staff  44578327 22 Mar  2016 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-1f869c5d2b5d94ddf1b46c35f4d48bc58d24ddc46aeb66327ce23f2f9a44fc69.zip
-rw-r--r--  1 felixsatyaputra  staff  44578324  7 Sep 17:55 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-9e348c46cb7eee755f932240ca4c3d8fb3cb17ebaa8c18c02b47caa5d294a443.zip
-rw-------  1 felixsatyaputra  staff  44578329 21 Oct 13:45 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-067414825020a56df569e948de18e7010fb05534d20e1b749760033652a572cd.zip
-rw-------  1 felixsatyaputra  staff  44578324  2 Nov 16:46 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-6a2f3414f4f9de39c78556aeae21952768f5a1a570bd3f446b7982567219b7b4.zip
```

Load latest `cassandra` from bintray. Bundle will be downloaded from Bintray and loaded into ConductR.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct load cassandra --api-version 1
Retrieving bundle...
Bintray credentials loaded from /Users/felixsatyaputra/.bintray/.credentials
Loading bundle from cache typesafe/bundle/cassandra
Retrieving file:///Users/felixsatyaputra/workspace/typesafe-fsat/conductr-cli/cassandra
Bintray credentials loaded from /Users/felixsatyaputra/.bintray/.credentials
Resolving bundle typesafe/bundle/cassandra
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/cassandra/v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
[#################################################] 100%
Loading bundle to ConductR...
[#################################################] 100%
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc waiting to be installed
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc installed
Bundle loaded.
Start bundle with: conduct run --api-version 1 4f71ca2
Unload bundle with: conduct unload --api-version 1 4f71ca2
Print ConductR info with: conduct info --api-version 1

Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct info
ID       NAME       #REP  #STR  #RUN
5a79de9  eslite        1     0     1
4f71ca2  cassandra     1     0     0
```

The older versions of Cassandra bundle is cleaned up. Only the latest version and the recently loaded Cassandra bundle is present in the cache.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ ls -lart ~/.conductr/cache/cassandra-v3-*
-rw-------  1 felixsatyaputra  staff  44578324  2 Nov 16:46 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-6a2f3414f4f9de39c78556aeae21952768f5a1a570bd3f446b7982567219b7b4.zip
-rw-------  1 felixsatyaputra  staff  44578324 14 Nov 18:04 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
```

Subsequent load works as well.

```
Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct load cassandra --api-version 1
Retrieving bundle...
Bintray credentials loaded from /Users/felixsatyaputra/.bintray/.credentials
Loading bundle from cache typesafe/bundle/cassandra
Retrieving from cache /Users/felixsatyaputra/.conductr/cache/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
Loading bundle to ConductR...
[#################################################] 100%
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc is installed
Bundle loaded.
Start bundle with: conduct run --api-version 1 4f71ca2
Unload bundle with: conduct unload --api-version 1 4f71ca2
Print ConductR info with: conduct info --api-version 1

Felixs-MBP-2:conductr-cli felixsatyaputra$ conduct info
ID       NAME       #REP  #STR  #RUN
5a79de9  eslite        1     0     1
4f71ca2  cassandra     1     0     0

Felixs-MBP-2:conductr-cli felixsatyaputra$ ls -lart ~/.conductr/cache/cassandra-v3-*
-rw-------  1 felixsatyaputra  staff  44578324  2 Nov 16:46 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-6a2f3414f4f9de39c78556aeae21952768f5a1a570bd3f446b7982567219b7b4.zip
-rw-------  1 felixsatyaputra  staff  44578324 14 Nov 18:04 /Users/felixsatyaputra/.conductr/cache/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
```

fsat commented:
![](https://github.trello.services/images/mini-trello-icon.png) [ConductR CLI: perform housekeeping for .cache directory](https://trello.com/c/s99AvoC1/382-conductr-cli-perform-housekeeping-for-cache-directory)

markusjura commented on conductr_cli/conduct_load.py:
```diff
@@ -223,3 +232,22 @@ def continue_logging(monitor):
+
+    bundle_files_to_delete = older_bundle_files[:(-1 * KEEP_BUNDLE_VERSIONS)]
+    for file in bundle_files_to_delete:
+        os.remove(file)
```
I suspect that `os.remove` does not work on Windows. We have a pending issue when deleting bundle configuration files: https://github.com/typesafehub/conductr-cli/issues/175

So before merging the PR we need to verify if the solution works on Windows as well.

fsat replied:
As tested by @markusjura on https://github.com/typesafehub/conductr-cli/pull/192#issuecomment-260288858 and myself, `os.remove` does work with files on Windows.

fsat commented:
Marked again as wip - need to test on Windows.

fsat commented:
**Manual test on Windows**

This is done by having ConductR listening on `192.168.99.100`. Ensure python on Windows is at latest (version `3.5.2`).

Checkout `conductr-cli` on the PR's branch.

```
IEUser@IE9Win7 MINGW32 ~/Workspace/conductr-cli ((3be0cab...))
$ pip3 install -e .
Obtaining file:///C:/Users/IEUser/Workspace/conductr-cli
Requirement already satisfied (use --upgrade to upgrade): requests>=2.6.0 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): requests-toolbelt>=0.7.0 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): argcomplete>=0.8.1 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pyhocon==0.2.1 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): arrow>=0.6.0 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): jsonschema==2.4 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pager<4.0,>=3.3 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): prettytable<1.0,>=0.7 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pygments<3.0,>=2.0 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): six<2.0,>=1.9 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): toml<1.0,>=0.9 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pyparsing==2.0.3 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from pyhocon==0.2.1->conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): python-dateutil in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from arrow>=0.6.0->conductr-cli==0.45)
Installing collected packages: conductr-cli
  Running setup.py develop for conductr-cli
Successfully installed conductr-cli-0.45
You are using pip version 8.1.1, however version 9.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.

IEUser@IE9Win7 MINGW32 ~/Workspace/conductr-cli ((3be0cab...))
$ conduct info --host 192.168.99.100
ID       NAME    #REP  #STR  #RUN
5a79de9  eslite     1     0     1
```

Prime the cache with older versions of the `cassandra` bundle.

```
IEUser@IE9Win7 MINGW32 ~/Workspace/conductr-cli ((3be0cab...))
$ ll ~/.conductr/cache/
total 224032
-rw-r--r-- 1 IEUser 197121 44578329 Oct 20 19:45 cassandra-v3-067414825020a56df569e948de18e7010fb05534d20e1b749760033652a572cd.zip
-rw-r--r-- 1 IEUser 197121 44578327 Mar 21  2016 cassandra-v3-1f869c5d2b5d94ddf1b46c35f4d48bc58d24ddc46aeb66327ce23f2f9a44fc69.zip
-rw-r--r-- 1 IEUser 197121 44578324 Nov  1 22:46 cassandra-v3-6a2f3414f4f9de39c78556aeae21952768f5a1a570bd3f446b7982567219b7b4.zip
-rw-r--r-- 1 IEUser 197121 44578324 Sep  7 00:55 cassandra-v3-9e348c46cb7eee755f932240ca4c3d8fb3cb17ebaa8c18c02b47caa5d294a443.zip
-rw-r--r-- 1 IEUser 197121 51083592 Mar 15  2016 reactive-maps-frontend-v1-a5eb1554d40d92d9cf3c11ba121f5624f6bc4c73b482dabb6ef6a545da168bbf.zip
```

Load the cassandra bundle.

```
IEUser@IE9Win7 MINGW32 ~/Workspace/conductr-cli ((3be0cab...))
$ conduct load cassandra --host 192.168.99.100
Retrieving bundle...
Cache directory C:\Users\IEUser/.conductr/cache has the permissions 777. Setting permissions to 700.
Loading bundle from cache typesafe/bundle/cassandra
Retrieving file:///C:\Users\IEUser\Workspace\conductr-cli\cassandra
Resolving bundle typesafe/bundle/cassandra
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/cassandra/v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f/cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
[#################################################] 100%
Loading bundle to ConductR...
[#################################################] 100%
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc waiting to be installed
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc installed
Bundle loaded.
Start bundle with: conduct run --host 192.168.99.100 4f71ca2
Unload bundle with: conduct unload --host 192.168.99.100 4f71ca2
Print ConductR info with: conduct info --host 192.168.99.100
```

The older versions of Cassandra bundle is cleaned up. Only the latest version and the recently loaded Cassandra bundle is present in the cache.

```
IEUser@IE9Win7 MINGW32 ~/Workspace/conductr-cli ((3be0cab...))
$ ll ~/.conductr/cache/
total 93424
-rw-r--r-- 1 IEUser 197121 44578324 Nov 13 23:36 cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
-rw-r--r-- 1 IEUser 197121 51083592 Mar 15  2016 reactive-maps-frontend-v1-a5eb1554d40d92d9cf3c11ba121f5624f6bc4c73b482dabb6ef6a545da168bbf.zip
```

fsat commented:
**Manual test on Windows (without Git Bash)**

It seems the oldest version that should be kept is not present after `conduct load`.

```
Microsoft Windows [Version 6.1.7601]
Copyright (c) 2009 Microsoft Corporation.  All rights reserved.

C:\Users\IEUser>python



C:\Users\IEUser\Workspace\conductr-cli>pip install -e .
Obtaining file:///C:/Users/IEUser/Workspace/conductr-cli
Requirement already satisfied (use --upgrade to upgrade): requests>=2.6.0 in c:\
users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from c
onductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): requests-toolbelt>=0.7
.0 in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-package
s (from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): argcomplete>=0.8.1 in
c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (fro
m conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pyhocon==0.2.1 in c:\u
sers\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from co
nductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): arrow>=0.6.0 in c:\use
rs\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from cond
uctr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): jsonschema==2.4 in c:\
users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from c
onductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pager<4.0,>=3.3 in c:\
users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from c
onductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): prettytable<1.0,>=0.7
in c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (
from conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pygments<3.0,>=2.0 in
c:\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (fro
m conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): six<2.0,>=1.9 in c:\us
ers\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from con
ductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): toml<1.0,>=0.9 in c:\u
sers\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from co
nductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): pyparsing==2.0.3 in c:
\users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from
pyhocon==0.2.1->conductr-cli==0.45)
Requirement already satisfied (use --upgrade to upgrade): python-dateutil in c:\
users\ieuser\appdata\local\programs\python\python35-32\lib\site-packages (from a
rrow>=0.6.0->conductr-cli==0.45)
Installing collected packages: conductr-cli
  Running setup.py develop for conductr-cli
Successfully installed conductr-cli-0.45
You are using pip version 8.1.1, however version 9.0.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' comm
and.

C:\Users\IEUser\Workspace\conductr-cli>conduct info --host 192.168.99.100
ID       NAME       #REP  #STR  #RUN
5a79de9  eslite        1     0     1
4f71ca2  cassandra     1     0     0

C:\Users\IEUser\Workspace\conductr-cli>conduct unload cass --host 192.168.99.100

Bundle unload request sent.
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc waiting to be uninstalled
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc uninstalled
Print ConductR info with: conduct info --host 192.168.99.100

C:\Users\IEUser\Workspace\conductr-cli>dir c:\users\IEUser\.conductr\cache
 Volume in drive C has no label.
 Volume Serial Number is 2431-AE02

 Directory of c:\users\IEUser\.conductr\cache

11/13/2016  11:43 PM    <DIR>          .
11/13/2016  11:43 PM    <DIR>          ..
10/20/2016  06:45 PM        44,578,329 cassandra-v3-067414825020a56df569e948de18
e7010fb05534d20e1b749760033652a572cd.zip
03/21/2016  04:42 PM        44,578,327 cassandra-v3-1f869c5d2b5d94ddf1b46c35f4d4
8bc58d24ddc46aeb66327ce23f2f9a44fc69.zip
11/01/2016  09:46 PM        44,578,324 cassandra-v3-6a2f3414f4f9de39c78556aeae21
952768f5a1a570bd3f446b7982567219b7b4.zip
09/06/2016  11:55 PM        44,578,324 cassandra-v3-9e348c46cb7eee755f932240ca4c
3d8fb3cb17ebaa8c18c02b47caa5d294a443.zip
03/15/2016  08:04 PM        51,083,592 reactive-maps-frontend-v1-a5eb1554d40d92d
9cf3c11ba121f5624f6bc4c73b482dabb6ef6a545da168bbf.zip
               5 File(s)    229,396,896 bytes
               2 Dir(s)  120,248,684,544 bytes free

C:\Users\IEUser\Workspace\conductr-cli>conduct load cassandra --host 192.168.99.
100
Retrieving bundle...
Cache directory C:\Users\IEUser/.conductr/cache has the permissions 777. Setting
 permissions to 700.
Loading bundle from cache typesafe/bundle/cassandra
Retrieving file:///C:\Users\IEUser\Workspace\conductr-cli\cassandra
Resolving bundle typesafe/bundle/cassandra
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/cassandra/v3-4f71ca2f
c8760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f/cassandra-v3-4f71ca2fc8
760ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
[#################################################] 100%
Loading bundle to ConductR...
[#################################################] 100%
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc waiting to be installed
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc installed
Bundle loaded.
Start bundle with: conduct run --host 192.168.99.100 4f71ca2
Unload bundle with: conduct unload --host 192.168.99.100 4f71ca2
Print ConductR info with: conduct info --host 192.168.99.100

C:\Users\IEUser\Workspace\conductr-cli>dir c:\users\IEUser\.conductr\cache
 Volume in drive C has no label.
 Volume Serial Number is 2431-AE02

 Directory of c:\users\IEUser\.conductr\cache

11/13/2016  11:45 PM    <DIR>          .
11/13/2016  11:45 PM    <DIR>          ..
11/13/2016  11:45 PM        44,578,324 cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfb
fbccc9c591cd8623b39858e0c4b9d24c465f.zip
03/15/2016  08:04 PM        51,083,592 reactive-maps-frontend-v1-a5eb1554d40d92d
9cf3c11ba121f5624f6bc4c73b482dabb6ef6a545da168bbf.zip
               2 File(s)     95,661,916 bytes
               2 Dir(s)  120,382,427,136 bytes free

C:\Users\IEUser\Workspace\conductr-cli>conduct load cassandra --host 192.168.99.
100
Retrieving bundle...
Cache directory C:\Users\IEUser/.conductr/cache has the permissions 777. Setting
 permissions to 700.
Loading bundle from cache typesafe/bundle/cassandra
Retrieving from cache C:\Users\IEUser/.conductr/cache/cassandra-v3-4f71ca2fc8760
ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
Loading bundle to ConductR...
[#################################################] 100%
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc is installed
Bundle loaded.
Start bundle with: conduct run --host 192.168.99.100 4f71ca2
Unload bundle with: conduct unload --host 192.168.99.100 4f71ca2
Print ConductR info with: conduct info --host 192.168.99.100

C:\Users\IEUser\Workspace\conductr-cli>
```

markusjura commented:
@fsat Did a quick sanity check on Windows as well. Used the `cmd` as a terminal window.

Deleting the file is working on Windows. 

However, that we do the deletion check based on the filename of the zip file. What if a user renames the filename of the same service `reactive-maps-frontend-v1`, e.g.:
- reactive-maps-frontend-v1-sha.zip
- reactive-maps-frontend-v2-sha.zip

In this case, we would identify the same service / bundle as two different services. What about using the `name` of the `bundle.conf` as an identifier?

Otherwise, looking good to me.

fsat commented:
Using `os.path.abspath` will yield correct, expected result when comparing files containing Windows paths.

As such, the Windows behaviour of the CLI when used with `CMD` is now correct.

```
C:\Users\IEUser\Workspace\conductr-cli>dir c:\users\IEUser\.conductr\cache
 Volume in drive C has no label.
 Volume Serial Number is 2431-AE02

 Directory of c:\users\IEUser\.conductr\cache

11/14/2016  03:39 PM    <DIR>          .
11/14/2016  03:39 PM    <DIR>          ..
10/20/2016  06:45 PM        44,578,329 cassandra-v3-067414825020a56df569e948de18
e7010fb05534d20e1b749760033652a572cd.zip
03/21/2016  04:42 PM        44,578,327 cassandra-v3-1f869c5d2b5d94ddf1b46c35f4d4
8bc58d24ddc46aeb66327ce23f2f9a44fc69.zip
11/14/2016  12:06 AM        44,578,324 cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfb
fbccc9c591cd8623b39858e0c4b9d24c465f.zip
11/01/2016  09:46 PM        44,578,324 cassandra-v3-6a2f3414f4f9de39c78556aeae21
952768f5a1a570bd3f446b7982567219b7b4.zip
09/06/2016  11:55 PM        44,578,324 cassandra-v3-9e348c46cb7eee755f932240ca4c
3d8fb3cb17ebaa8c18c02b47caa5d294a443.zip
03/15/2016  08:04 PM        51,083,592 reactive-maps-frontend-v1-a5eb1554d40d92d
9cf3c11ba121f5624f6bc4c73b482dabb6ef6a545da168bbf.zip
               6 File(s)    273,975,220 bytes
               2 Dir(s)  119,924,162,560 bytes free

C:\Users\IEUser\Workspace\conductr-cli>conduct load cassandra --host 192.168.99.
100
Retrieving bundle...
Cache directory C:\Users\IEUser/.conductr/cache has the permissions 777. Setting
 permissions to 700.
Loading bundle from cache typesafe/bundle/cassandra
Retrieving from cache C:\Users\IEUser/.conductr/cache/cassandra-v3-4f71ca2fc8760
ddabadfd8bdcdfbfbccc9c591cd8623b39858e0c4b9d24c465f.zip
Loading bundle to ConductR...
[#################################################] 100%
Bundle 4f71ca2fc8760ddabadfd8bdcdfbfbcc is installed
Bundle loaded.
Start bundle with: conduct run --host 192.168.99.100 4f71ca2
Unload bundle with: conduct unload --host 192.168.99.100 4f71ca2
Print ConductR info with: conduct info --host 192.168.99.100

C:\Users\IEUser\Workspace\conductr-cli>dir c:\users\IEUser\.conductr\cache
 Volume in drive C has no label.
 Volume Serial Number is 2431-AE02

 Directory of c:\users\IEUser\.conductr\cache

11/14/2016  03:40 PM    <DIR>          .
11/14/2016  03:40 PM    <DIR>          ..
11/14/2016  12:06 AM        44,578,324 cassandra-v3-4f71ca2fc8760ddabadfd8bdcdfb
fbccc9c591cd8623b39858e0c4b9d24c465f.zip
11/01/2016  09:46 PM        44,578,324 cassandra-v3-6a2f3414f4f9de39c78556aeae21
952768f5a1a570bd3f446b7982567219b7b4.zip
03/15/2016  08:04 PM        51,083,592 reactive-maps-frontend-v1-a5eb1554d40d92d
9cf3c11ba121f5624f6bc4c73b482dabb6ef6a545da168bbf.zip
               3 File(s)    140,240,240 bytes
               2 Dir(s)  120,057,909,248 bytes free


C:\Users\IEUser\Workspace\conductr-cli>conduct info --host 192.168.99.100
ID       NAME       #REP  #STR  #RUN
5a79de9  eslite        1     0     1
4f71ca2  cassandra     1     0     0
```

fsat commented:
Still marked as wip - will retest scenario other than Windows.

fsat commented:
Weird build failure due to formatting error reported by `flake8` - didn't get such error when running `python3 setup.py test` locally.

```
========================== 296 passed in 1.96 seconds ==========================
flake8 create: /home/travis/build/typesafehub/conductr-cli/.tox/flake8
flake8 installdeps: flake8, pep8-naming
flake8 inst: /home/travis/build/typesafehub/conductr-cli/.tox/dist/conductr-cli-0.46.zip
flake8 installed: argcomplete==1.6.0,arrow==0.8.0,conductr-cli==0.46,flake8==3.2.0,jsonschema==2.4.0,mccabe==0.5.2,pager==3.3,pep8-naming==0.4.1,prettytable==0.7.2,pycodestyle==2.2.0,pyflakes==1.3.0,Pygments==2.1.3,pyhocon==0.2.1,pyparsing==2.0.3,python-dateutil==2.6.0,requests==2.11.1,requests-toolbelt==0.7.0,six==1.10.0,toml==0.9.2
flake8 runtests: PYTHONHASHSEED='2538034910'
flake8 runtests: commands[0] | flake8 --ignore=E501 .
./setup.py:59:1: E305 expected 2 blank lines after class or function definition, found 1
./conductr_cli/conduct.py:12:1: E305 expected 2 blank lines after class or function definition, found 1
./conductr_cli/sandbox.py:12:1: E305 expected 2 blank lines after class or function definition, found 1
./conductr_cli/shazar.py:12:1: E305 expected 2 blank lines after class or function definition, found 1
./dcos/subcommand.py:394:1: E305 expected 2 blank lines after class or function definition, found 1
ERROR: InvocationError: '/home/travis/build/typesafehub/conductr-cli/.tox/flake8/bin/flake8 --ignore=E501 .'
```

The error all occurs in the files not touched by this PR. Not sure why this is happening now.

Will try rebasing with master.

fsat committed:
Fix formatting error reported by flake8

Not sure why this is only reported now - these errors has been there all along.

fsat commented:
Repeated the following manual tests. Everything is working as expected.
- `Manual test with Conduct Load using V2 of the Control Protocol`
- `Manual test with Conduct Load using V1 of the Control Protocol`
